src/control/lib/netdetect/hwloc/hwloc.go

```diff
@@ -0,0 +1,133 @@
+//
+// (C) Copyright 2021 Intel Corporation.
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+//
+
+// hwloc is a set of Go bindings for interacting with the hwloc library.
+package hwloc
+
+import (
+	"github.com/pkg/errors"
+)
+
+// API is an interface for basic API operations, including synchronizing access.
+type API interface {
```

Comment:
Another quick thought... This isn't wrong, per se, but this use of interfaces isn't really idiomatic. It's not java. ;) In Go, interfaces are most usefully defined by the user, i.e. "This is what I need the thing to do in this method -- as long as it can do this, I don't care what else it can do." The standard library is full of examples like this, e.g. io.Writer, etc.

That's why I'm suggesting that we put the interface definitions in the netdetect package. It will dictate its requirements for its methods, and then we can make the hwloc/libfabric/ucx/etc implementations adhere to those requirements.

Reply:
I ended up defining the interfaces here because I wanted the hwloc getters to be able to return interface types instead of concrete types. Seemed a bit more testable. Maybe I just need to shift all of the interfaces up into netdetect and restructure.

Reply:
I'm not strictly opposed to using interfaces that way, but the general advice for idiomatic Go is to accept interfaces and return concrete types. This is in keeping with the philosophy that the consumer/user should be defining the functionality it wants, rather than a type defining the functionality it provides.

Mostly I just want to shift the focus on testing to our business logic rather than the underlying sources of information, and I do agree that this will likely require some more restructuring.